File: org.dcs.core/src/main/java/org/dcs/core/module/flow/TestFlowModule.java
package org.dcs.core.module.flow;

import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;

import org.dcs.api.model.TestResponse;
import org.dcs.api.service.FlowModuleConstants;
import org.dcs.api.service.RESTException;
import org.dcs.api.service.TestApiService;
import org.osgi.framework.BundleContext;
import org.osgi.framework.ServiceReference;

public class TestFlowModule implements FlowModule {

	
	private TestApiService testService;

	
  public static final Map<String, Properties> properties = new HashMap<>();
  
  public static final Map<String, Properties> relationships = new HashMap<>();
  
  

  public static final String PROPERTY_USER_NAME_VALUE = "User Name";

	@Override
	public void init(BundleContext bundleContext) {		
    ServiceReference reference = bundleContext.getServiceReference(TestApiService.class.getName());
    testService = (TestApiService) bundleContext.getService(reference);
    
		Properties userNameProperties = new Properties();
		userNameProperties.put(FlowModuleConstants.PROPERTY_NAME, PROPERTY_USER_NAME_VALUE);
		userNameProperties.put(FlowModuleConstants.PROPERTY_DESCRIPTION, "User To Greet");
		userNameProperties.put(FlowModuleConstants.PROPERTY_REQUIRED, false);
		userNameProperties.put(FlowModuleConstants.PROPERTY_DEFAULT_VALUE, "");
		
		properties.put(PROPERTY_USER_NAME_VALUE, userNameProperties);
		
		Properties userNameRelationships = new Properties();
		userNameRelationships.put(FlowModuleConstants.PROPERTY_NAME, FlowModuleConstants.REL_SUCCESS_ID);
		userNameRelationships.put(FlowModuleConstants.PROPERTY_DESCRIPTION, "All status updates will be routed to this relationship");
		
		relationships.put(FlowModuleConstants.REL_SUCCESS_ID, userNameRelationships);
	}
	
	@Override
	public Map<String, Properties> getPropertyDescriptors() {
    return properties;
	}

	@Override
	public Map<String, Properties> getRelationships() {
    return relationships;
	}
	
	@Override
	public void schedule() {		
		
	}

	@Override
	public byte[] trigger(Properties properties) throws RESTException {		
		TestResponse response= testService.testHelloGet(properties.getProperty(PROPERTY_USER_NAME_VALUE));
		return response.getResponse().getBytes(StandardCharsets.UTF_8);
	}

	@Override
	public void unschedule() {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void stop() {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void shutdown() {
		// TODO Auto-generated method stub
		
	}

	@Override
	public void remove() {
		// TODO Auto-generated method stub
		
	}

}

File: org.dcs.core/src/main/scala/org/dcs/core/module/flow/TestFlowModule.scala
package org.dcs.core.module.flow

import java.nio.charset.StandardCharsets
import java.util.{Map => JavaMap}

import org.dcs.api.service.{FlowModule, TestApiService, TestResponse}
import org.osgi.framework.{BundleContext, ServiceReference}

import scala.collection.JavaConverters._
import org.dcs.commons.JsonSerializerImplicits._

object TestFlowModule {
  val PropertyUserNameValue = "User Name";
}

class TestFlowModule extends FlowModule {

  var testService: TestApiService = _

  val propertiesMap = {
    val userNameProperties = collection.mutable.Map(
      FlowModule.PropertyName -> TestFlowModule.PropertyUserNameValue,
      FlowModule.PropertyDescription -> "User To Greet",
      FlowModule.PropertyRequired -> "false",
      FlowModule.PropertyDefaultValue -> "")

    collection.mutable.Map(TestFlowModule.PropertyUserNameValue -> userNameProperties.asJava).asJava
  }
  //    val userNameProperties = new HashMap[String, String]()
  //  userNameProperties.put(FlowModule.PropertyName, TestFlowModule.PropertyUserNameValue)
  //  userNameProperties.put(FlowModule.PropertyDescription, "User To Greet")
  //  userNameProperties.put(FlowModule.PropertyRequired, "false")
  //  userNameProperties.put(FlowModule.PropertyDefaultValue, "")
  //
  //  val propertiesMap = new HashMap[String, JavaMap[String, String]]()
  //  propertiesMap.put(TestFlowModule.PropertyUserNameValue, userNameProperties)

  val relationshipsMap = {
    val userNameRelationships = collection.mutable.Map(
      FlowModule.PropertyName -> FlowModule.RelSuccessId,
      FlowModule.PropertyDescription -> "All status updates will be routed to this relationship")

    collection.mutable.Map(FlowModule.RelSuccessId -> userNameRelationships.asJava).asJava
  }
  //    val userNameRelationships = new HashMap[String, String]()
  //  userNameRelationships.put(FlowModule.PropertyName, FlowModule.RelSuccessId)
  //  userNameRelationships.put(FlowModule.PropertyDescription, "All status updates will be routed to this relationship")
  //
  //  val relationshipsMap = new HashMap[String, JavaMap[String, String]]()
  //  relationshipsMap.put(FlowModule.RelSuccessId, userNameRelationships)

  override def init(bundleContext: BundleContext) {
    val className = classOf[TestApiService].getName
    val reference: ServiceReference[TestApiService] =
      bundleContext.getServiceReference[TestApiService](classOf[TestApiService])
    testService = bundleContext.getService(reference);

  }

  override def getPropertyDescriptors(): JavaMap[String, JavaMap[String, String]] = {
    propertiesMap;
  }

  override def getRelationships(): JavaMap[String, JavaMap[String, String]] = {
    relationshipsMap;
  }

  override def schedule() {

  }

  override def trigger(properties: JavaMap[String, String]): Array[Byte] = {
    testService.
      hello(properties.get(TestFlowModule.PropertyUserNameValue)).
      toJson.
      getBytes(StandardCharsets.UTF_8)
  }

  override def unschedule() {

  }

  override def stop() {

  }

  override def shutdown() {

  }

  override def remove() {

  }

}